core/src/main/scala/org/apache/spark/io/MutableCheckedOutputStream.scala

```diff
@@ -0,0 +1,47 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.io
+
+import java.io.OutputStream
+import java.util.zip.Checksum
+
+/**
+ * A variant of [[java.util.zip.CheckedOutputStream]] which can
+ * change the checksum calculator at runtime.
+ */
+class MutableCheckedOutputStream(out: OutputStream) extends OutputStream {
+  private var checksum: Checksum = _
+
+  def setChecksum(c: Checksum): Unit = this.checksum = c
```

Comment:
> Here and elsewhere, prevent changing checksum once set ? (assert((this.checksum eq null) || (checksum eq c)) )

Actually, this is intended to allow changing the `checksum` instance at runtime.  (If you look at the comment of `MutableCheckedOutputStream`, it's also mentioned.)

It's because, in the case of `UnsafeShuffleWriter`, there could be multiple spill files. And, each spill file could generate data for multiple partitions. Thus, during the writing of each spill file, we should change `checksum` instance for different partitions. Besides, the lifetime of `checksum` instances should cover all the spill files so that they can calculate the checksums for the completed partitions. That's why we define the global `partitionChecksums` in `ShuffleExternalSorter`.

Reply:
Thanks for clarifying !